File: Loom.Providers/OpenAI/ModelResponse.vb
Namespace OpenAI
    Public Class ModelResponse
        Public Property Id As String
        Public Property [Object] As String
        Public Property Created_at As Long
        Public Property Status As String
        Public Property [Error] As Object
        Public Property Incomplete_details As Object
        Public Property Instructions As String
        Public Property Max_output_tokens As Object
        Public Property Model As String
        Public Property Output As List(Of OutputItem)
        Public Property Parallel_tool_calls As Boolean
        Public Property Previous_response_id As String
        Public Property Reasoning As Reasoning
        Public Property Store As Boolean
        Public Property Temperature As Double
        Public Property Text As TextConfig
        Public Property Tool_choice As String
        Public Property Tools As List(Of Tool)
        Public Property Top_p As Double
        Public Property Truncation As String
        Public Property Usage As Usage
        Public Property User As Object
        Public Property Metadata As Dictionary(Of String, Object)
    End Class
    Public Class OutputItem
        Public Property Id As String
        Public Property Type As String
        Public Property Status As String
        Public Property Arguments As String
        Public Property Call_id As String
        Public Property Name As String
        Public Property Queries As List(Of String)
        Public Property Results As Object
        Public Property Role As String
        Public Property Content As List(Of MessageContent)
    End Class
    Public Class MessageContent
        Public Property Type As String
        Public Property Text As String
        Public Property Annotations As List(Of Annotation)
    End Class
    Public Class Annotation
        Public Property Type As String
        Public Property Index As Integer
        Public Property File_id As String
        Public Property Filename As String
    End Class
    Public Class Reasoning
        Public Property Effort As Object
        Public Property Summary As Object
    End Class
    Public Class TextConfig
        Public Property Format As FormatConfig
        Public Property Verbosity As String
    End Class
    Public Class FormatConfig
        Public Property Type As String
    End Class
    Public Class Tool
        Public Property Type As String
        Public Property Description As String
        Public Property Name As String
        Public Property Parameters As Dictionary(Of String, Object)
        Public Property Strict As Boolean
        Public Property Filters As Object
        Public Property Max_num_results As Integer?
        Public Property Ranking_options As Dictionary(Of String, Object)
        Public Property Vector_store_ids As List(Of String)
    End Class
    Public Class Usage
        Public Property Input_tokens As Integer
        Public Property Input_tokens_details As InputTokensDetails
        Public Property Output_tokens As Integer
        Public Property Output_tokens_details As OutputTokensDetails
        Public Property Total_tokens As Integer
    End Class
    Public Class InputTokensDetails
        Public Property Cached_tokens As Integer
    End Class
    Public Class OutputTokensDetails
        Public Property Reasoning_tokens As Integer
    End Class
End Namespace

File: implementations/csharp/src/Loom.Providers.OpenAI/ModelResponse.cs
using System.Collections.Generic;

namespace Loom.Providers.OpenAI
{
    public class ModelResponse
    {
        public string Id { get; set; }
        public string Object { get; set; }
        public long Created_at { get; set; }
        public string Status { get; set; }
        public object Error { get; set; }
        public object Incomplete_details { get; set; }
        public string Instructions { get; set; }
        public object Max_output_tokens { get; set; }
        public string Model { get; set; }
        public List<OutputItem> Output { get; set; }
        public bool Parallel_tool_calls { get; set; }
        public string Previous_response_id { get; set; }
        public Reasoning Reasoning { get; set; }
        public bool Store { get; set; }
        public double Temperature { get; set; }
        public TextConfig Text { get; set; }
        public string Tool_choice { get; set; }
        public List<Tool> Tools { get; set; }
        public double Top_p { get; set; }
        public string Truncation { get; set; }
        public Usage Usage { get; set; }
        public object User { get; set; }
        public Dictionary<string, object> Metadata { get; set; }
    }

    public class OutputItem
    {
        public string Id { get; set; }
        public string Type { get; set; }
        public string Status { get; set; }
        public string Arguments { get; set; }
        public string Call_id { get; set; }
        public string Name { get; set; }
        public List<string> Queries { get; set; }
        public object Results { get; set; }
        public string Role { get; set; }
        public List<MessageContent> Content { get; set; }
    }

    public class MessageContent
    {
        public string Type { get; set; }
        public string Text { get; set; }
        public List<Annotation> Annotations { get; set; }
    }

    public class Annotation
    {
        public string Type { get; set; }
        public int Index { get; set; }
        public string File_id { get; set; }
        public string Filename { get; set; }
    }

    public class Reasoning
    {
        public object Effort { get; set; }
        public object Summary { get; set; }
    }

    public class TextConfig
    {
        public FormatConfig Format { get; set; }
        public string Verbosity { get; set; }
    }

    public class FormatConfig
    {
        public string Type { get; set; }
    }

    public class Tool
    {
        public string Type { get; set; }
        public string Description { get; set; }
        public string Name { get; set; }
        public Dictionary<string, object> Parameters { get; set; }
        public bool Strict { get; set; }
        public object Filters { get; set; }
        public int? Max_num_results { get; set; }
        public Dictionary<string, object> Ranking_options { get; set; }
        public List<string> Vector_store_ids { get; set; }
    }

    public class Usage
    {
        public int Input_tokens { get; set; }
        public InputTokensDetails Input_tokens_details { get; set; }
        public int Output_tokens { get; set; }
        public OutputTokensDetails Output_tokens_details { get; set; }
        public int Total_tokens { get; set; }
    }

    public class InputTokensDetails
    {
        public int Cached_tokens { get; set; }
    }

    public class OutputTokensDetails
    {
        public int Reasoning_tokens { get; set; }
    }
}
